### PR TITLE
Add thermostat tube module

### PR DESCRIPTION
**Pull request by johalun**
This change adds a thermostat tube module. It measures the temperature of the block it's facing (including air), and sends a redstone signal of 15 on the selected color channel when temperature is above set threshold for the simple version. When upgraded it can be configured with lower and higher bounds with an interpolated redstone output, similar to the pressure gauge module.

Added

- Thermostat module and related code, recipe, textures, etc.
- `AbstractTubeModule:canConnectTo` function that is used by `ModuleNetworkManager:computeConnections`. The default implementation is to only returns objects of same class to keep the previous behavior, but tube modules can override it to be included in for example calculation of redstone signals output by redstone modules which the thermostat module does.
- `WidgetColorSelector.isExpanded` public getter. This is used to disable the bounds sliders when the color selector is expanded in the thermostat advanced GUI.
 
Changed

- ~~The function `ModuleNetworkManager.java:computeConnections` only returned modules of the same class. This is called by the redstone module set to output when iterating over all connected modules to get the highest redstone signal. To make it work with the new class I removed this filter (it needs to also iterate over thermostat modules now). Anywhere where this function is called, the caller also filters on class type so there's no difference in functionality, expect some performance impact since we now return more objects than before that the caller has to filter through. If this is a concern we could maybe add a function parameter for the caller to supply a list of valid class types to be returned.~~ 

Todo

- [X] Add advanced mode. Maybe similar to pressure gauge module?
- [ ] Make a unique 3D model and texture? I don't have the software needed to edit 3D objects and export texture atlases so I simply modified the texture of the redstone module and used its 3D model. Maybe would be nice to have a bit of a different 3D model for this module? Any recommendations for software and tools to use to make the 3D models?
- [X] Clean up code
- [X] Add entry in manual
- [X] Improve tooltip description

### CODE REVIEW

**Comment by desht** [comment]
Thanks for this.
* I think a filter function listing what can see what would be good, to limit the number of modules returned
* We have an official texture artist who might do a model if asked nicely, @Ridanisaurus ? :smiley:

File: src/main/java/me/desht/pneumaticcraft/common/tubemodules/ModuleNetworkManager.java [hunk 1]
```diff
@@ -60,7 +60,7 @@ private Set<AbstractTubeModule> computeConnections(AbstractTubeModule module) {
         while (!pendingPositions.isEmpty()) {
             BlockPos pos = pendingPositions.pop();
             PneumaticCraftUtils.getTileEntityAt(level, pos, PressureTubeBlockEntity.class).ifPresent(tube -> tube.tubeModules()
-                    .filter(tm -> tm instanceof INetworkedModule && module.getClass() == tm.getClass())
+                    .filter(tm -> tm instanceof INetworkedModule)
```
Comment (desht) [inline src/main/java/me/desht/pneumaticcraft/common/tubemodules/ModuleNetworkManager.java]: Might be worth adding an `AbstractTubeModule#canConnectTo(AbstractTubeModule other)` method for this.  Default would be to check for the same class, but could be overridden for the thermostat module to also connect to redstone modules.

Comment (johalun) [reply src/main/java/me/desht/pneumaticcraft/common/tubemodules/ModuleNetworkManager.java]: That's sounds like a good idea.

**Comment by johalun** [comment]
> We have an official texture artist

Yes please :) I couldn't make it nearly as good.

**Comment by johalun** [comment]
@desht Do you have a nifty way to auto join a test world when starting the client with `./gradlew runClient`?  Edit: I tried https://github.com/florensie/auto-load but couldn't get it to work. Maybe out of date?..

**Comment by johalun** [comment]
@desht Please give it a test run now. Turned out quite ok I think. Hope I didn't miss anything obvious. Just the new 3D model missing.

**Comment by johalun** [comment]
Regarding the recipe. I was thinking it would be fitting if we could make one ingredient be something heat related. Maybe heat pipe or a manometer or something similar. What do you think?

**Comment by desht** [comment]
Yeah something including both, perhaps?

(Didn't get the chance to test the changes over the weekend, sorry - soon, hopefully!)

**Comment by Ridanisaurus** [comment]
I'll work on a model this weekend.

**Comment by desht** [comment]
I got a chance to play with this, and it works well, although I have a fairly minor point to raise.

* If you set the threshold to (say) 50C, the signal doesn't switch from 0 -> 15 until the temperature is 51C. Maybe that's OK, but I think it switching on at 50C would be more intuitive.
* Similarly, using the advanced interface, I set the signal to interpolate between 0C and 200C. At 200C the redstone output is 14, not 15 (at 201C the signal is 15). Again, I think it would be more intuitive to have the upper limit of the temperature range correspond to the upper range of the output signal?

File: src/main/java/me/desht/pneumaticcraft/common/tubemodules/ThermostatModule.java [hunk 2]
```diff
@@ -156,6 +173,11 @@ public void onNeighborTileUpdate() {
         updateInputLevel();
     }
 
+    @Override
+    public boolean canConnectTo(AbstractTubeModule other) {
+        return true;
```
Comment (desht) [inline src/main/java/me/desht/pneumaticcraft/common/tubemodules/ThermostatModule.java:179]: This should really be `other instanceof RedstoneModule || super.canConnectTo(other)`. There's no reason for thermostat modules to connect to any other module, is there?

**Comment by johalun** [comment]
> I got a chance to play with this, and it works well, although I have a fairly minor point to raise.
> 
>     * If you set the threshold to (say) 50C, the signal doesn't switch from 0 -> 15 until the temperature is 51C. Maybe that's OK, but I think it switching on at 50C would be more intuitive.
> 
>     * Similarly, using the advanced interface, I set the signal to interpolate between 0C and 200C. At 200C the redstone output is 14, not 15 (at 201C the signal is 15). Again, I think it would be more intuitive to have the upper limit of the temperature range correspond to the upper range of the output signal?

Yes I agree. It's an oversight on my part.

File: src/main/java/me/desht/pneumaticcraft/common/tubemodules/ThermostatModule.java (hunk 2 above)
Comment (johalun) [reply src/main/java/me/desht/pneumaticcraft/common/tubemodules/ThermostatModule.java:179]: Good point!

**Comment by johalun** [comment]
> If you set the threshold to (say) 50C, the signal doesn't switch from 0 -> 15 until the temperature is 51C. Maybe that's OK, but I think it switching on at 50C would be more intuitive.

@desht  Did you test with an older commit? I changed half way through from `>` to `>=` for the simple case. I also noticed that the creative compressed iron block lag behind in updating neighboring blocks so when you increase the temp step by step, the thermostat is always one step behind, i.e. compressed iron 49->50, thermostat reads 49, next step from 50->51, thermostat reads 50, and so on. Could that be a factor here?

https://github.com/TeamPneumatic/pnc-repressurized/pull/1264/files#diff-04804b2120d213f9a4e48bbd369de9bb05d8646b882497145956fd41b2b319bfR207

Edit: same for the advanced. I'm getting 200C -> 15 redstone. But I have to make sure that the thermostat is actually reading the current temp of the compressed iron block by forcing a block update. Could this be an issue in the thermostat code or how the compressed iron block is doing neighbor block update on changes?

**Comment by desht** [comment]
It's possible, I'll check again later 🙂

**Comment by Ridanisaurus** [comment]
thermostat_module_3
Model is pretty much done, just need to create the upgraded texture and work on the render.

**Comment by johalun** [comment]
@Ridanisaurus That looks great!

**Comment by johalun** [comment]
@desht Do you think we can also get this into 1.19.2? Can't wait to start using it :)

**Comment by desht** [comment]
Yeah, sorry for the delay, just been too busy... should have some time next week though. I have a nice model from Rid, but it needs some manual patching to wrangle it into the code, so I will merge this into dev and patch the model in with a separate commit.

Hopefully some time shortly after the 25th :christmas_tree:

**Comment by johalun** [comment]
@desht Thanks for merging! When do you think we'll have release with this included for 1.19.2? Trying to coordinate with the Valhelsia team to get this in the next mod pack update.

**Comment by desht** [comment]
This week, with luck... Thu/Fri perhaps?

**Pull request merged**